File: src/optic/lua/flat/ops/Call.java
package optic.lua.flat.ops;

import optic.lua.flat.*;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class Call implements Step {
	private final StepType type;
	private final Register function;
	private final Register output;
	private final List<Register> args;

	public Call(StepType type, Register function, List<Register> args) {
		this(type, function, args, Register.unused());
	}

	public Call(StepType call, Register function, List<Register> args, Register output) {
		type = call;
		this.function = function;
		this.output = output;
		this.args = args;
	}

	@NotNull
	@Override
	public StepType getType() {
		return type;
	}

	@Override
	public String toString() {
		boolean resultUnused = output.equals(Register.unused());
		return resultUnused
				? String.format("%s %s(%s)", typeName(), function, args)
				: String.format("%s %s = %s(%s)", typeName(), output, function, args);
	}
}

File: src/optic/lua/asm/instructions/Call.kt
package optic.lua.asm.instructions

import optic.lua.asm.RValue
import optic.lua.asm.Register
import optic.lua.asm.Step
import java.util.function.Consumer

@Deprecated("Use Invoke in future")
class Call(val function: Register, val args: List<Register>, val output: Register) : Step {
    override fun toString(): String {
        return if (output.isUnused())
            "call $function($args)"
        else
            "call $output = $function($args)"
    }

    override fun modified(): Register? {
        return when {
            output.isUnused() -> null
            else -> output
        }
    }

    override fun forEachObserved(action: Consumer<RValue>) {
        args.forEach(action)
        action.accept(function)
    }
}
